qubes-rpc/qubes.DesktopNotify

```diff
@@ -0,0 +1,320 @@
+    global BUS
+    if not BUS:
+        BUS = dbus.SessionBus()
+    proxy = BUS.get_object(BUS_NAME, BUS_PATH)
+    return dbus.Interface(proxy, BUS_NAME)
+
+def _notify(trusted_notification):
+    iface = get_iface() #we shouldn't cache it as we'll run into issues otherwise (cf. notification-forwarder)
+
+    return iface.Notify(
+        trusted_notification.app_name,
+        trusted_notification.notification_id,
+        trusted_notification.app_icon,
+        trusted_notification.summary,
+        trusted_notification.body,
+        trusted_notification.actions,
+        trusted_notification.hints,
+        trusted_notification.expire_timeout,
+        signature="susssasa{sv}i"
+    )
+
+def notify(vm, trusted_notification):
+    notification = trusted_notification
+
+    #the ID is the caller ID --> fix it
+    our_id = IDS.get(notification.notification_id, 0)
+
+    #give the sending VM a dedicated field (summary) to hopefully avoid spoofing attacks (e.g. with backspaces)
+    summary = vm
+    if notification.body and notification.summary:
+        body = " : ".join([notification.summary, notification.body])
+    elif notification.body:
+        body = notification.body
+    else:
+        body = notification.summary
+
+    notification = Notification(
+        app_name=notification.app_name,
+        notification_id=our_id,
+        app_icon=notification.app_icon,
+        summary=summary,
+        body=body,
+        actions=notification.actions,
+        hints=notification.hints,
+        expire_timeout=notification.expire_timeout,
+    )
+
+    LOG.info(f"Notification from VM {vm}: {notification}")
+    ret_id = _notify(notification)
+
+    if trusted_notification.notification_id > 0:
+        IDS[trusted_notification.notification_id] = ret_id
+
+def close_notification(trusted_id):
+    #the ID is the caller ID --> fix it
+    try:
+        our_id = IDS.pop(trusted_id)
+    except KeyError:
+        LOG.warning("Received a closed request for a notification that's not in our databse. This is unexpected. Ignoring...")
+        return
+
+    iface = get_iface()
+    iface.CloseNotification(our_id, signature="u")
+
+def regex_sanitize(vm, regex, untrusted_str, field):
+    m = re.match(regex, untrusted_str)
+    if m:
+        return m.group(1)
+    raise ParseException(f"Received unexpected message content in the field {field} from VM {vm}.")
+
+def parse_notification_id(vm, untrusted):
+    return int(regex_sanitize(vm, r"^([0-9]{1,6})$", untrusted, "notification_id"))
+
+def parse_expire_timeout(vm, untrusted):
+    return int(regex_sanitize(vm, r"^(\-?[0-9]{1,10})$", untrusted, "expire_timeout"))
+
+def assert_printable(vm, untrusted, max_len):
+    l = len(untrusted)
+    if l >= max_len:
+        raise ParseException(f"Received content of excessive length ({l} >= {max_len}) from VM {vm}.")
+
+    allowed_chars = set(["\t", "\n", "\r"])
+    ret = []
+    for ch in untrusted:
+        cat = unicodedata.category(ch)
```

Comment:
Then I'll leave that to whoever will write that renderer. ;-) 

One simply cannot foresee and prevent all bugs in all possible viewers by filtering further upstream.

Reply:
That is true, but I would like to think there is a middle ground between “ASCII only” and “full Unicode” that can at least include Marek Markzykowski-Górecki’s last name.

Reply:
Agreed, but I do believe there is a middle ground between “ASCII only” and “full Unicode”.  For instance, this comment contains non-ASCII characters, but I do not believe it is significantly more dangerous than one that is pure ASCII.

That said, finding such a middle ground is beyond the scope of this PR, and would lead to changes elsewhere in Qubes OS.  At the very least, the GUI daemon should be changed, too.